File: src/ClaudeToolbar.Core/Usage/UsageClient.cs
using System.Net;
using System.Net.Http.Headers;
using ClaudeToolbar.Core.Time;

namespace ClaudeToolbar.Core.Usage;

public sealed class UsageClient : IUsageClient
{
    public const string Endpoint = "https://api.anthropic.com/api/oauth/usage";
    public const string UserAgent = "claude-code/2.0.0";
    public const string BetaHeader = "oauth-2025-04-20";
    public static readonly TimeSpan RequestTimeout = TimeSpan.FromSeconds(10);

    private readonly HttpClient _http;
    private readonly IClock _clock;

    public UsageClient(HttpClient http, IClock clock)
    {
        _http = http;
        _clock = clock;
    }

    public async Task<UsageResult> FetchAsync(string accessToken, CancellationToken cancellationToken)
    {
        using var request = new HttpRequestMessage(HttpMethod.Get, Endpoint);
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", accessToken);
        request.Headers.TryAddWithoutValidation("anthropic-beta", BetaHeader);
        request.Headers.TryAddWithoutValidation("User-Agent", UserAgent);
        request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));

        using var timeoutCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
        timeoutCts.CancelAfter(RequestTimeout);

        try
        {
            using var response = await _http.SendAsync(request, HttpCompletionOption.ResponseHeadersRead, timeoutCts.Token).ConfigureAwait(false);

            if (response.StatusCode is HttpStatusCode.Unauthorized or HttpStatusCode.Forbidden)
                return new UsageResult.Unauthorized();

            if ((int)response.StatusCode == 429)
                return new UsageResult.RateLimited(ReadRetryAfter(response));

            if (!response.IsSuccessStatusCode)
                return new UsageResult.Failed($"HTTP {(int)response.StatusCode}");

            var json = await response.Content.ReadAsStringAsync(timeoutCts.Token).ConfigureAwait(false);
            return UsageResponseParser.Parse(json, _clock.UtcNow);
        }
        catch (OperationCanceledException) when (!cancellationToken.IsCancellationRequested)
        {
            return new UsageResult.Failed("Request timed out");
        }
        catch (HttpRequestException ex)
        {
            return new UsageResult.Failed(ex.Message);
        }
        catch (IOException ex)
        {
            return new UsageResult.Failed(ex.Message);
        }
    }

    private TimeSpan? ReadRetryAfter(HttpResponseMessage response)
    {
        var header = response.Headers.RetryAfter;
        if (header is null) return null;
        if (header.Delta is { } delta) return delta;
        if (header.Date is { } date) return date - _clock.UtcNow;
        return null;
    }
}

File: mac/Sources/ClaudeToolbarCore/Usage/UsageClient.swift
import Foundation
#if canImport(FoundationNetworking)
import FoundationNetworking
#endif

public protocol UsageClient: Sendable {
    func fetch(accessToken: String) async -> UsageResult
}

public struct OAuthUsageClient: UsageClient {
    public static let endpoint = URL(string: "https://api.anthropic.com/api/oauth/usage")!
    public static let userAgent = "claude-code/2.0.0"
    public static let betaHeader = "oauth-2025-04-20"
    public static let requestTimeout: TimeInterval = 10

    private let transport: any HTTPTransport
    private let clock: any ClockSource

    public init(transport: any HTTPTransport, clock: any ClockSource) {
        self.transport = transport
        self.clock = clock
    }

    public func fetch(accessToken: String) async -> UsageResult {
        var request = URLRequest(url: Self.endpoint, cachePolicy: .reloadIgnoringLocalCacheData, timeoutInterval: Self.requestTimeout)
        request.httpMethod = "GET"
        request.setValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")
        request.setValue(Self.betaHeader, forHTTPHeaderField: "anthropic-beta")
        request.setValue(Self.userAgent, forHTTPHeaderField: "User-Agent")
        request.setValue("application/json", forHTTPHeaderField: "Accept")

        let response: HTTPResponse
        do {
            response = try await transport.send(request)
        } catch {
            return .failed(Self.describe(error))
        }

        switch response.statusCode {
        case 401, 403:
            return .unauthorized
        case 429:
            return .rateLimited(retryAfter: Self.retryAfter(response.headers["retry-after"], now: clock.now))
        case 200..<300:
            return UsageResponseParser.parse(String(decoding: response.body, as: UTF8.self), fetchedAt: clock.now)
        default:
            return .failed("HTTP \(response.statusCode)")
        }
    }

    static func retryAfter(_ header: String?, now: Date) -> TimeInterval? {
        guard let header = header?.trimmingCharacters(in: .whitespaces), !header.isEmpty else { return nil }
        if let seconds = Double(header) { return max(seconds, 0) }
        if let date = HTTPDate.parse(header) { return date.timeIntervalSince(now) }
        return nil
    }

    static func describe(_ error: Error) -> String {
        if let urlError = error as? URLError, urlError.code == .timedOut { return "Request timed out" }
        return error.localizedDescription
    }
}
